Support multi root paths

Poytr1 opened the pull request:
Resolve: https://github.com/elastic/code/issues/1360

movie-travel-code commented on src/lsp-server.ts:
```diff
@@ -49,15 +50,17 @@ export class LspServer {
     didChangeWorkspaceFolders(params: DidChangeWorkspaceFoldersParams) {
-        this.rootPath = fileURLToPath(params.event.added[0].uri);
-        this.runCtags(this.rootPath);
+        const rootPath = fileURLToPath(params.event.added[0].uri);
```
I have question that if we delete a repository will code server send `didChangeWorkspaceFolder` request with the removed folders?

We just extract the first element of `params.event.added` here, is there a chance omit remain folders. In other way, is it a strong gurantee that `parasm.event.added` just has one element?

Poytr1 replied:
Yeh, I missed that, I will add the removed workspace logic.

Poytr1 replied:
Done.

Poytr1 commented:
Yeh, I missed that, I will add the removed workspace logic.

On Fri, Jun 21, 2019 at 11:43 AM Henry Wong <notifications@github.com>
wrote:

> *@movie-travel-code* commented on this pull request.
> ------------------------------
>
> In src/lsp-server.ts
> <https://github.com/elastic/ctags-langserver/pull/8#discussion_r296086167>
> :
>
> > @@ -49,15 +50,17 @@ export class LspServer {
>      }
>
>      didChangeWorkspaceFolders(params: DidChangeWorkspaceFoldersParams) {
> -        this.rootPath = fileURLToPath(params.event.added[0].uri);
> -        this.runCtags(this.rootPath);
> +        const rootPath = fileURLToPath(params.event.added[0].uri);
>
> I have question that if we delete a repository will code server send
> didChangeWorkspaceFolder request with the removed folders?
>
> We just extract the first element of params.event.added here, is there a
> chance omit remain folders. In other way, is it a strong gurantee that
> parasm.event.added just has one element?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/ctags-langserver/pull/8?email_source=notifications&email_token=AE4VLPUR5F63KCZWXEBMPTLP3RE75A5CNFSM4HZRDLU2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB4HM76Y#pullrequestreview-252628987>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE4VLPSXASCRXWPIIZ3LJITP3RE75ANCNFSM4HZRDLUQ>
> .
>


-- 
*Pengcheng Xu*
Software Engineer - APJ
pengcheng.xu@elastic.co |  www.elastic.co

movie-travel-code commented on src/lsp-server.ts:
```diff
@@ -49,15 +50,27 @@ export class LspServer {
+        removed.forEach(remove => {
+            const index = this.rootPaths.indexOf(fileURLToPath(remove.uri));
+            if (index !== -1) {
+                this.rootPaths.slice(index, 1);
```
I'm not a ts expert. Do you mean `splice` here? `slice` method will return a new array, the original array will not be modified, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice.

Poytr1 replied:
I should use `splice` instead of `slice` actually. And It's done now
